cql3/expr/prepare_expr.cc

```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
     return ::make_shared<assignment_testable_expression>(std::move(e));
 }
 
+// Finds column_defintion for given column name in the schema.
+// Second argument is a relation to print in the error message in case of failure.
+static const column_definition* resolve_column(const unresolved_identifier& col_ident, const schema& schema, const binary_operator& printable_relation) {
+    ::shared_ptr<column_identifier> id = col_ident.ident->prepare_column_identifier(schema);
+    const column_definition* def = get_column_definition(schema, *id);
+    if (!def || def->is_hidden_from_cql()) {
+        throw exceptions::unrecognized_entity_exception(*id, fmt::format("{}", printable_relation));
+    }
+    return def;
+}
+
+// Resolves columns on LHS of binary_operator and prepares the index in case of a subscripted column.
+// Last argument is a relation to print in the error message in case of failure.
+static expression prepare_binop_lhs(const expression& lhs, data_dictionary::database db, const schema& schema, const binary_operator& printable_relation) {
+    return expr::visit(overloaded_functor{
+        [&](const unresolved_identifier& unin) -> expression {
+            return column_value(resolve_column(unin, schema, printable_relation));
+        },
+        [](const column_value& cv) -> expression { return cv; },
+        [&](const tuple_constructor& tc) -> expression {
+            std::vector<expression> new_elements;
+            new_elements.reserve(tc.elements.size());
+
+            for (const expression& elem : tc.elements) {
+                new_elements.emplace_back(prepare_binop_lhs(elem, db, schema, printable_relation));
+            }
+
+            return tuple_constructor {
+                .elements = std::move(new_elements)
+            };
+        },
+        [&](const token& tk) -> expression {
+            std::vector<expression> prepared_token_args;
+            prepared_token_args.reserve(tk.args.size());
+
+            for (const expression& arg : tk.args) {
+                prepared_token_args.emplace_back(prepare_binop_lhs(arg, db, schema, printable_relation));
+            }
+
+            return token(std::move(prepared_token_args));
+        },
+        [&](const subscript& sub) -> expression {
+            const column_definition* sub_col = expr::visit(overloaded_functor {
+                [&](const unresolved_identifier& unin) -> const column_definition* {
+                    return resolve_column(unin, schema, printable_relation);
+                },
```

Comment:
And here too, better to call prepare_expression recursively than to create new logic.

Reply:
I think we need to add a `data_type` field to expressions. prepare_expression will fill this type.

Reply:
Well, we have it in a weak form in get_lhs_receiver(). It can form the basis of more generalization later.

Reply:
`resolve_column` requires `schema`, which isn't passed as an argument to `prepare_expression`, but could probably be added.